pallets/services/src/lib.rs

```diff
@@ -1099,5 +1054,59 @@ pub mod module {
 			// TODO: add weight for the call to the total weight.
 			Ok(PostDispatchInfo { actual_weight: None, pays_fee: Pays::Yes })
 		}
+
+		pub fn slash(
+			origin: OriginFor<T>,
+			offender: T::AccountId,
+			#[pallet::compact] service_id: u64,
+			#[pallet::compact] percent: Percent,
+		) -> DispatchResult {
+			let caller = ensure_signed(origin)?;
+			let service = Self::services(service_id)?;
+			// TODO: add EVM hook to query for the slashing permission.
+			ensure!(service.owner == caller, DispatchError::BadOrigin);
+			let service_operator =
+				service.operators.iter().find(|(operator, _)| operator == &offender);
+			let (operator, restake_percent) = match service_operator {
+				Some((operator, restake_percent)) => (operator, restake_percent),
+				None => return Err(Error::<T>::OffenderNotOperator.into()),
+			};
+			// sanity check
+			ensure!(operator == &offender, Error::<T>::OffenderNotOperator);
```

Comment:
This doesn't seem needed. You literally filtered for it above.

Reply:
Seems even that lines 1068 and the code block on 1070 can be turned into 1 call.